sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedPlainValuesReader.java

```diff
@@ -53,19 +53,50 @@ public void skip() {
     throw new UnsupportedOperationException();
   }
 
+  private void updateCurrentByte() {
+    try {
+      currentByte = (byte) in.read();
+    } catch (IOException e) {
+      throw new ParquetDecodingException("Failed to read a byte", e);
+    }
+  }
+
   @Override
   public final void readBooleans(int total, WritableColumnVector c, int rowId) {
-    // TODO: properly vectorize this
-    for (int i = 0; i < total; i++) {
-      c.putBoolean(rowId + i, readBoolean());
+    int i = 0;
+    if (bitOffset > 0) {
+      i = Math.min(8 - bitOffset, total);
+      c.putBooleans(rowId, i, currentByte, bitOffset);
+      bitOffset = (bitOffset + i) & 7;
+    }
+    for (; i + 7 < total; i += 8) {
+      updateCurrentByte();
+      c.putBooleans(rowId + i, currentByte);
+    }
+    if (i < total) {
+      updateCurrentByte();
+      bitOffset = total - i;
+      c.putBooleans(rowId + i, bitOffset, currentByte, 0);
     }
   }
 
   @Override
   public final void skipBooleans(int total) {
-    // TODO: properly vectorize this
-    for (int i = 0; i < total; i++) {
-      readBoolean();
+    // Using >>3 instead of /8 below. The difference is important when (total-(8-bitOffset))<0.
+    // E.g. (-1)>>3=(-1) vs. (-1)/8=0. The latter incorrectly enters the if(numBytesToSkip>=0){.
+    // (total-(8-bitOffset))<0 means there will be still unread bits left in the currentByte.
+    // In that case, updateCurrentByte() should not be called.
+    int numBytesToSkip = (total - (8 - bitOffset)) >> 3;
+    bitOffset = (bitOffset + total) & 7;
+    if (numBytesToSkip >= 0) {
```

Comment:
what if I first skip N bits where N < 8, and then read some bits? e.g.,
```java
        reader.skipBooleans(4)
        reader.readBooleans(4, vector, 0)
```

it seems we won't update the `currentByte` and the read will read incorrect values?

Reply:
Thank you @sunchao for finding this! I missed it. Since this logic is error prone and I didn't like the long comments, I changed the method to follow the same structure as `readBooleans()`. I have also updated the test case to cover this case.